Added parameter CALICO_NAT_OUTGOING for default IPPool Creation

VincentS opened the pull request:
This patch relates to the issue: #1580

It adds the parameter to CALICO_NAT_OUTGOING for the default IPPool creation.
This enables to disable nat-outgoing in the default IPv4 and IPv6 IPPools created.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/calicoctl?pullRequest=1640) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/calicoctl?pullRequest=1640) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/calicoctl?pullRequest=1640) it.</sub>

tmjd commented:
This could use some tests too.  Take a look at #1471 for the tests that were added for the other default pool configuration variables.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -524,11 +526,26 @@ func ipv6Supported() bool {
+// default IPPool Creation. It returns a boolean.
+
+func defineNATOutgoing () bool {
+	switch strings.ToLower(os.Getenv("CALICO_NAT_OUTGOING")) {
```
I think this name should be `CALICO_IPPOOL_NAT_OUTGOING` if there is only one variable for both IPv4 and IPv6.  If, as I suspect, that we should have 2 then it should match the format of the other default pool configuration variables.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -524,11 +526,26 @@ func ipv6Supported() bool {
+// defineNATOutgoing checks the CALICO_NAT_OUTGOING parameter used for
+// default IPPool Creation. It returns a boolean.
+
+func defineNATOutgoing () bool {
```
I think this should be named something like `isNATOutgoingEnabled`

VincentS commented:
Refactored the code based on the feedback. I introduced a new helper function called `evaluateENVBool` to prevent code duplication (2 - 3 identical almost switch statements). 
With this refactoring it would be quite simple to separate NAT_OUTGOING configuration for IPv4 and IPv6.

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,21 @@ func validateIP(ipn *net.IPNet) {
+
+// evaluateENVBool evaluates a passed environment variable and writes an
+// log entry if false otherwise it returns true by default
+
```
nit: could you remove the extra newline here?

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -450,6 +465,7 @@ func configureIPPools(client *client.Client) {
 		return
 	}
 
+
```
nit: and this one too :)

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,21 @@ func validateIP(ipn *net.IPNet) {
+	}
+
+	return true
+
```
nit: extra newline

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -497,23 +513,22 @@ func configureIPPools(client *client.Client) {
 	if !ipv6Present && ipv6Supported() {
 		log.Debug("Create default IPv6 IP pool")
-		createIPPool(client, ipv6Cidr, string(ipip.Undefined))
+		createIPPool(client, ipv6Cidr, string(ipip.Undefined), evaluateENVBool("CALICO_IPPOOL_NAT_OUTGOING","NAT_Outgoing disabled for Default IPv6 Pool."))
```
@fasaxc does Felix actually support outgoing NAT on ipv6 pools?

VincentS replied:
This is still TBD

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -497,23 +513,22 @@ func configureIPPools(client *client.Client) {
-	case "false", "0", "no", "n", "f":
-		log.Info("IPv6 support disabled through environment")
+	// First check if Felix param is false
+	if !evaluateENVBool("FELIX_IPV6SUPPORT","IPv6 support disabled through environment") {
```
nit: could you add spaces after each argument?

e.g.

```
("FELIX_IPV6SUPPORT", "IPv6 support disabled through environment")
```

VincentS commented:
I made the changes based on the feedback by @caseydavenport. Test will be added soon.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -497,23 +509,22 @@ func configureIPPools(client *client.Client) {
 	if !ipv4Present {
 		log.Debug("Create default IPv4 IP pool")
-		createIPPool(client, ipv4Cidr, ipv4IpipModeEnvVar)
+		createIPPool(client, ipv4Cidr, ipv4IpipModeEnvVar, evaluateENVBool("CALICO_IPPOOL_NAT_OUTGOING", "NAT_Outgoing disabled for Default IPv4 Pool."))
```
Please change the environment variable names to be IP version specific.
ie. CALICO_IPV4POOL_NAT_OUTGOING

tmjd commented:
@caseydavenport Can you confirm that we do want a separate IPv4 and IPv6 env var for IPIP?

caseydavenport commented:
@tmjd @VincentS sorry, yes I think we do (although NAT outgoing on IPv6 pools is probably not generally something people _should_ be doing, so that flag should default to `False`)

tmjd commented:
@VincentS It has been a few days and I wanted to check in with you, is there anything I can help with on this PR or an idea of when you may have more updates?

VincentS commented:
Hey I havent been in the office last week.

VincentS commented:
I adjusted and rebased/squashed the commits of the PR. Currently Tests are still WIP since some of the new are failing at the moment.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -526,10 +544,16 @@ func ipv6Supported() bool {
 
+
+	//Workaround for NatOutgoing where IPv6 default is false
+	if(version == 6 && isNATOutgoingDefined == false) {
```
I think this should be placed in configureIPPools, there is already an IPv6 section there and then it won't be necessary to pass isNATOutgoingDefined into createIPPool.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,22 @@ func validateIP(ipn *net.IPNet) {
+// and an boolean if the argument was true or false
+// If it was false an additional log entry will be written
+
+func evaluateENVBool(envvar string, logentry string) (isDefined bool, isValue bool ) {
```
I think there could be a better name for the 'isValue' variable.  Maybe envValue or possibly just value, or the extra verbose option envBoolValue.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -497,23 +513,25 @@ func configureIPPools(client *client.Client) {
 	if !ipv4Present {
 		log.Debug("Create default IPv4 IP pool")
-		createIPPool(client, ipv4Cidr, ipv4IpipModeEnvVar)
+		OutgoingNATisDefined, OutgoingNATisValue := evaluateENVBool("CALICO_IPV4POOL_NAT_OUTGOING", "NAT_Outgoing disabled for Default IPv4 Pool.")
```
With other suggested changes I think OutgoingNATisValue should just be outgoingNATEnabled.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -497,23 +513,25 @@ func configureIPPools(client *client.Client) {
-	case "false", "0", "no", "n", "f":
-		log.Info("IPv6 support disabled through environment")
+	// First check if Felix param is false
+	_, isValue := evaluateENVBool("FELIX_IPV6SUPPORT", "IPv6 support disabled through environment");
```
I think whatever you change isValue to in evaluateENVBool you can probably use that name here too.

tmjd commented on calico_node/startup/startup_test.go:
```diff
@@ -137,31 +152,49 @@ var _ = Describe("FV tests against a real etcd", func() {
+		Entry("IPv6 NATOutgoing Set Enabled",
+			[]EnvItem{
+				{"CALICO_IPV6POOL_NAT_OUTGOING", "true"}},
+			"192.168.0.0/16", "fdff:ffff:ffff:ffff:ffff::/80", "off", true, true),
```
You need to use the default IPv6 pool here. `fd80:24e2:f998:72d6::/64` like in the first test entry

tmjd commented on calico_node/startup/startup_test.go:
```diff
@@ -137,31 +152,49 @@ var _ = Describe("FV tests against a real etcd", func() {
+		Entry("IPv6 NATOutgoing Set Disabled",
+			[]EnvItem{
+				{"CALICO_IPV6POOL_NAT_OUTGOING", "false"}},
+			"192.168.0.0/16", "fdff:ffff:ffff:ffff:ffff::/80", "off", true, false),
```
Same with the IPv6 default here and for the 2 test Entries below too

tmjd commented on calico_node/startup/startup_test.go:
```diff
@@ -137,31 +152,49 @@ var _ = Describe("FV tests against a real etcd", func() {
+				{"CALICO_IPV4POOL_NAT_OUTGOING", "false"},
+				{"CALICO_IPV6POOL_NAT_OUTGOING", "true"},
+			},
+			"192.168.0.0/16", "fdff:ffff:ffff:ffff:ffff::/80", "always", false, true),
```
`"always"` will require having CALICO_IPV4POOL_IPIP set.

VincentS commented:
Fixed all remaining issues and squashed / rebase commit.

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,22 @@ func validateIP(ipn *net.IPNet) {
+// Returns an boolean indicating if the parameter was defined
+// and an boolean if the argument was true or false
+// If it was false an additional log entry will be written
+
```
Could you remove this extra blank line here?

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,22 @@ func validateIP(ipn *net.IPNet) {
+// and an boolean if the argument was true or false
+// If it was false an additional log entry will be written
+
+func evaluateENVBool(envvar string, logentry string) (isDefined bool, outgoingNATEnabled bool) {
```
I think the return values here shouldn't be named unless they're being used within the function.

Generally, we don't use named returns in Calico code, so I think it would be best to remove the names and ensure the code is commented with correct info.

caseydavenport replied:
nit: `envvar` should be named `envVar`

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,22 @@ func validateIP(ipn *net.IPNet) {
+	envValue, isEmpty := os.LookupEnv(envvar)
+	switch strings.ToLower(envValue) {
+	case "false", "0", "no", "n", "f":
+		log.Info(logentry)
```
I think the logging would be cleaner if instead of passing in the log entry, we just logged something like this:

```
log.Info("%s (%s) is disabled through environment variable", envVar, envValue)
```

Then we won't need to pass in the logs throughout, and all the logs will be guaranteed to have the same format.

tmjd replied:
I think we should always log when the environment variable is set.  @caseydavenport ??

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -497,11 +513,19 @@ func configureIPPools(client *client.Client) {
-		createIPPool(client, ipv6Cidr, string(ipip.Undefined))
+		outgoingNATisDefined, outgoingNATEnabled := evaluateENVBool("CALICO_IPV6POOL_NAT_OUTGOING", "NAT_Outgoing disabled for Default IPv6 Pool.")
+
+		//Workaround for NatOutgoing where IPv6 default is false
```
nit: needs a space after the slashes. `// Workaround`

caseydavenport commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,22 @@ func validateIP(ipn *net.IPNet) {
+		return isEmpty, false
+	}
+
+	return isEmpty, true
```
I think we should tweak this just a little bit.  IIUC, right now the code will return `true` even if the environment variable isn't defined.  I think a nicer interface would be that a default is passed in, and if the environment variable isn't defined the default is returned.

e.g.

```
// Returns True if the envVar is defined and set to true.
// Returns False if the envVar is defined and set to false.
// Returns def in the envVar is not defined.
func evaluateENVBool(envVar string, def bool) (bool, bool)
```

WDYT?

It would remove the need for this bit of code below:

```
if outgoingNATisDefined == false {
    outgoingNATEnabled = false
}
```

tmjd replied:
@caseydavenport Your example still shows 2 returned values though now we could get by with just one, do you think the isDefined return value should remain? I don't think so.

I had considered suggesting a change like this but had thought it allowed more flexibility because then the calling function could take a specific action if it was not defined.  That flexibility isn't really necessary here so I'm good with the change.  Plus with Casey's suggestion the feature to return if it is defined or not could always be added back.

caseydavenport replied:
> do you think the isDefined return value should remain? I don't think so.

Yeah, I missed that - we don't need the isDefined return value if we take this approach.  If we need an isDefined down the road for some other bit of code, we can add that (or that code can check it explicitly).

VincentS replied:
Not sure If I understood your suggestion correctly in my current PR.

tmjd replied:
@VincentS and I chatted in slack to clarify the request.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -526,7 +549,7 @@ func ipv6Supported() bool {
 // createIPPool creates an IP pool using the specified CIDR.  This
 // method is a no-op if the pool already exists.
-func createIPPool(client *client.Client, cidr *net.IPNet, ipipModeName string) {
+func createIPPool(client *client.Client, cidr *net.IPNet, ipipModeName string, isNATOutgoingDefined bool, isNATOutgoingEnabled bool) {
```
isNATOutgoingDefined is not used in the function anymore, please remove it.

tmjd commented on calico_node/startup/startup_test.go:
```diff
@@ -128,6 +136,13 @@ var _ = Describe("FV tests against a real etcd", func() {
 						Expect(pool.Spec.IPIP.Mode).To(Equal(ipip.Mode(expectIpv4IpipMode)))
 					}
+
+					if expectedIPV4NATOutgoing == false {
```
I think you should have the same logic as the IPV6 and not check that it `== false` and switch the 2 cases below.

tmjd replied:
I think it would be better to replace this if statement and the nested checks with
`Expect(pool.Spec.NATOutgoing).To(Equal(expectedIPv6NATOutgoing))`

tmjd commented:
I just looked at the failing tests and it looks like there are some of the python tests that were checking the defaults on IPV6 pools are failing because of the change to the default behavior.
Can you fix those please?
@caseydavenport We'll need to ensure we have a note about the change in default behavior in the release notes.  How do we do that?

tmjd commented:
@caseydavenport Do you think it is useful to add a new parameterized python test that uses these new env vars?  Kind of overkill but it is probably just a copy paste and a few modifications.

tmjd commented on calico_node/startup/startup_test.go:
```diff
@@ -118,6 +119,13 @@ var _ = Describe("FV tests against a real etcd", func() {
 						Expect(pool.Spec.IPIP.Enabled).To(BeFalse())
 					}
+
+					if expectedIPV6NATOutgoing {
```
I think it would be better to replace this if statement and the nested checks with
`Expect(pool.Spec.NATOutgoing).To(Equal(expectedIPv4NATOutgoing))`

caseydavenport commented:
@tmjd @VincentS I'd really like to get this guy merged if we can, what's left to do here?

VincentS commented:
Sorry, some other things came up and I didnt had time to finish it up. I begun to do some of the requested changes and hope to get the rest done by this week.

VincentS commented:
All changes were incorporated based on the feedback.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -510,10 +535,9 @@ func configureIPPools(client *client.Client) {
-	case "false", "0", "no", "n", "f":
-		log.Info("IPv6 support disabled through environment")
+	// First check if Felix param is false
+	outgoingNATEnabled := evaluateENVBool("FELIX_IPV6SUPPORT", false)
```
This outgoingNATEnabled is not an appropriate name.
I don't think the default was false previously.  If FELIX_IPV6SUPPORT was not set then I think the value would have been the empty string which would not match one of those cases.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -341,6 +341,28 @@ func validateIP(ipn *net.IPNet) {
+func evaluateENVBool(envVar string, defaultValue bool) bool {
+	envValue, isEmpty := os.LookupEnv(envVar)
+
+	log.Info("%s is %s through environment variable", envVar, envValue)
```
I think you should log the value that this function is returning, that way if the user is setting the value incorrectly it will be obvious if the variable is not being interpreted as they expect.
I think it would be nice to also include "(defaulted)" in the log message if the default value is being used.

You need to use `log.Infof` to make a formatted message, otherwise you get `%s is %s through environment variableCALICO_IPV4POOL_NAT_OUTGOING` as the output

tmjd commented on calico_node/startup/startup_test.go:
```diff
@@ -118,6 +119,9 @@ var _ = Describe("FV tests against a real etcd", func() {
 						Expect(pool.Spec.IPIP.Enabled).To(BeFalse())
 					}
+
+					Expect(pool.Spec.NATOutgoing).To(Equal(expectedIPV6NATOutgoing))
```
I think you can change that line to this and it will produce a much better error. `Expect(pool.Spec.NATOutgoing).To(Equal(expectedIPV6NATOutgoing), "Expected IPv6 to be %t but was %t", expectedIPV6NATOutgoing, pool.Spec.NATOutgoing)`

tmjd commented on calico_node/startup/startup_test.go:
```diff
@@ -128,6 +132,9 @@ var _ = Describe("FV tests against a real etcd", func() {
 						Expect(pool.Spec.IPIP.Mode).To(Equal(ipip.Mode(expectIpv4IpipMode)))
 					}
+
+					Expect(pool.Spec.NATOutgoing).To(Equal(expectedIPV4NATOutgoing))
```
Similar change as with the IPv6 version.

tmjd commented on calico_node/startup/startup.go:
```diff
@@ -526,7 +550,7 @@ func ipv6Supported() bool {
 // createIPPool creates an IP pool using the specified CIDR.  This
 // method is a no-op if the pool already exists.
-func createIPPool(client *client.Client, cidr *net.IPNet, ipipModeName string) {
+func createIPPool(client *client.Client, cidr *net.IPNet, ipipModeName string, isNATOutgoingEnabled bool) {
```
Please update the `message` below that has `with NAT outgoing enabled` to use the passed in isNATOutgoingEnabled.  You can use %t to convert a bool into the string `true` or `false`.

VincentS commented:
Close PR because it was moved to the Calico Repository https://github.com/projectcalico/calico/pull/861
